### PR TITLE
Adds mech calls to trades.py

### PR DESCRIPTION
**Pull request by jmoreira-valory**
Adds mech calls to trades.py

The scripts updates a local JSON database where it collects Mech events. It examines the two Mech contracts (the old and the new), therefore it needs to update the database twice.

The database update process can be cancelled at any time (Ctrl-C), but the remainder of the script will not have the data available to compute Mech costs.

### CODE REVIEW

File: trades.py [hunk 1]
```diff
@@ -174,47 +180,38 @@ def argparse(s: str) -> "MarketAttribute":
 STATS_TABLE_ROWS = list(MarketAttribute)
 
 
-def _read_file(file_path):
-    with open(file_path, 'r') as file:
-        return file.readline().strip()
-
-
-def _get_balance(address, rpc_url):
+def _get_balance(address: str, rpc_url: str) -> int:
```
Comment (jmoreira-valory) [inline trades.py:183]: This method is not part of this PR. Re-formatted by linter tool.

File: scripts/mech_events.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    # Old Mech contract
+    (
+        to_checksum_address("0xff82123dfb52ab75c417195c5fdb87630145ae81"),
+        28911547,
+    ),
+    # New Mech contract
+    (
+        to_checksum_address("0x77af31de935740567cf4ff1986d04b2c964a786a"),
+        30663133,
+    ),
```
Comment (0xArdi) [inline scripts/mech_events.py:63]: The Deliver events in these two contracts have different signatures. You will need both ABIs to get Delivers from these two contracts. This is how its done in the mech:
https://github.com/valory-xyz/mech/blob/main/packages/valory/contracts/agent_mech/contract.py#L201

File: scripts/mech_events.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        contract_instance = w3.eth.contract(address=mech_contract_address, abi=abi)
+
+        last_processed_block = (
+            mech_events_data.get(sender, {})
+            .get(mech_contract_address, {})
+            .get(event_name, {})
+            .get("last_processed_block", 0)
+        )
+        starting_block = max(earliest_block, last_processed_block + 1)
+        ending_block = w3.eth.get_block(LATEST_BLOCK_NAME)[BLOCK_DATA_NUMBER]
```
Comment (0xArdi) [inline scripts/mech_events.py:190]: Make sure this range is not >5K blocks, otherwise it will either timeout or error out.

File: scripts/mech_events.py (hunk 2 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py:63]: We are not retrieving Deliver events in this PR.

File: scripts/mech_events.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            contract_data = json.load(file)
+
+        abi = contract_data.get("abi", [])
+        contract_instance = w3.eth.contract(address=mech_contract_address, abi=abi)
+
+        last_processed_block = (
+            mech_events_data.get(sender, {})
+            .get(mech_contract_address, {})
+            .get(event_name, {})
+            .get("last_processed_block", 0)
```
Comment (0xArdi) [inline scripts/mech_events.py:187]: Maybe it makes sense to use the block from which the contracts were deployed.

File: scripts/mech_events.py (hunk 2 above)
Comment (0xArdi) [reply scripts/mech_events.py:63]: Okay the `MechDeliver` dataclass gave me the impression its used somewhere.

File: scripts/mech_events.py (hunk 3 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py:190]: 'Ending block' is the end of range block, not the "to_block" in the filter.

Comment (jmoreira-valory) [reply scripts/mech_events.py:190]: `            to_block = min(from_block + BLOCKS_CHUNK_SIZE, ending_block)
`

Here is the condition you were asking.

File: scripts/mech_events.py (hunk 4 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py:187]: This must be changed in lines 57 and 62.

**Review by 0xArdi — APPROVED** [review]

File: scripts/mech_events.py (hunk 4 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py:187]: Fixed

File: scripts/mech_events.py (hunk 2 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py]: Removed the class MechDeliver. If needed in future PRs can be re-added, and the issue you raised addressed.

File: scripts/mech_events.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        super().__init__(
+            event_id=args["requestId"],
+            data=args["data"].hex(),
+            sender=args["sender"],
+            transaction_hash=event["transactionHash"].hex(),
+            block_number=event["blockNumber"],
+        )
+
+        self.request_id = self.event_id
+        self.fee = 10000000000000000
```
Comment (jmoreira-valory) [inline scripts/mech_events.py:132]: This is the default value of the fees right now. Should be updated in a futue PR as the fee sent to the Mech contract address on the transaction associated to the event (we have the tx_hash).

Comment (DavidMinarsch) [reply scripts/mech_events.py:132]: Add a comment in code then please

File: trades.py [hunk 6]
```diff
@@ -21,16 +21,20 @@
...
+from pathlib import Path
 from string import Template
-from typing import Any
+from typing import Any, Dict, List
 
 import requests
 
+from scripts.mech_events import MechRequest, get_mech_requests
 
+IRRELEVANT_TOOLS = ["openai-text-davinci-002", "openai-text-davinci-003", "openai-gpt-3.5-turbo", "openai-gpt-4", "stabilityai-stable-diffusion-v1-5", "stabilityai-stable-diffusion-xl-beta-v2-2-2", "stabilityai-stable-diffusion-512-v2-1", "stabilityai-stable-diffusion-768-v2-1", "deepmind-optimization-strong", "deepmind-optimization"]
```
Comment (DavidMinarsch) [inline trades.py:37]: This is stored on-chain. Should be fetched to avoid having to manually update it.

File: scripts/mech_events.py (hunk 5 above)
Comment (jmoreira-valory) [reply scripts/mech_events.py:132]: Added.

File: trades.py (hunk 6 above)
Comment (jmoreira-valory) [reply trades.py:37]: I believe this is not stored on-chain. Probably you refer to available tools, but in this case is the tools the trader is excluding from the available ones.

@Adamantios please correct me if I'm wrong.
